src/commands/multi_block/monitor.rs

```diff
@@ -22,9 +22,117 @@ use polkadot_sdk::{
     pallet_election_provider_multi_block::unsigned::miner::MinerConfig,
     sp_npos_elections::ElectionScore,
 };
-use std::{collections::HashSet, sync::Arc};
+use std::{
+    collections::{HashMap, HashSet},
+    sync::Arc,
+};
 use tokio::sync::Mutex;
 
+/// Enum representing different tracking events for rounds
+#[derive(Debug, Clone, Copy)]
+enum RoundTrackingEvent {
+    Found,            // Found existing submission on-chain
+    Submitted,        // Successfully submitted a new solution
+    SubmittedPartial, // Submitted missing pages for partial solution
+    Cleared,          // Successfully cleared a discarded submission
+    FailedSubmission, // Failed to submit a solution
+}
+
+/// Enum representing different untracking events for rounds
+#[derive(Debug, Clone, Copy)]
+enum RoundUntrackingEvent {
+    Cleared, // Removed after clearing
+    Bailed,  // Removed after bailing
+    Failed,  // Removed after a failed submission
+}
+
+/// A dedicated type to manage submitted rounds tracking with clean APIs
+/// that hide the mutex implementation details
+#[derive(Clone)]
+struct RoundSubmission(Arc<Mutex<HashMap<u32, (u32, ElectionScore)>>>);
+
+impl RoundSubmission {
+    /// Create a new RoundSubmission tracker
+    fn new() -> Self {
+        Self(Arc::new(Mutex::new(HashMap::new())))
+    }
+
+    /// Check if a round is being tracked
+    async fn contains(&self, round: u32) -> bool {
+        self.0.lock().await.contains_key(&round)
+    }
+
+    /// Insert a new round submission into the tracking map
+    async fn insert(
+        &self,
+        round: u32,
+        n_pages: u32,
+        score: ElectionScore,
+        event: Option<RoundTrackingEvent>,
+    ) {
+        let mut rounds = self.0.lock().await;
+        rounds.insert(round, (n_pages, score));
+
+        if let Some(event) = event {
+            let message = match event {
```

Comment:
not that important, but you could make this a display impl on the type itself to make method impl slightly cleaner.

```rust
impl std::fmt::Display for RoundTrackingEvent {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            RoundTrackingEvent::Found => write!(f, "Found submission for"),
            RoundTrackingEvent::Submitted => write!(f, "Successfully submitted and recorded""),
            RoundTrackingEvent::SubmittedPartial => write!(f, "Completed partial submission for"),
            RoundTrackingEvent::Cleared => write!(f, "Cleared discarded submission for"),
            RoundTrackingEvent::FailedSubmission => write!(f, "Failed submission for"),
        }
    }
}
```